presto-hive/src/test/java/com/facebook/presto/hive/parquet/write/TestDataWritableWriter.java

```diff
@@ -0,0 +1,378 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.parquet.write;
+
+import io.airlift.log.Logger;
+import org.apache.hadoop.hive.common.type.HiveDecimal;
+import org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe;
+import org.apache.hadoop.hive.ql.io.parquet.timestamp.NanoTimeUtils;
+import org.apache.hadoop.hive.ql.io.parquet.write.DataWritableWriter;
+import org.apache.hadoop.hive.serde2.io.DateWritable;
+import org.apache.hadoop.hive.serde2.io.ParquetHiveRecord;
+import org.apache.hadoop.hive.serde2.objectinspector.ListObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.MapObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.BinaryObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.BooleanObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.ByteObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.DateObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.DoubleObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.FloatObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.HiveCharObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.HiveVarcharObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.IntObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.LongObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.ShortObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.StringObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.TimestampObjectInspector;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import parquet.io.api.Binary;
+import parquet.io.api.RecordConsumer;
+import parquet.schema.GroupType;
+import parquet.schema.OriginalType;
+import parquet.schema.Type;
+
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Copy of org.apache.hadoop.hive.ql.io.parquet.write.DataWritableWriter with patch for empty array and empty map
+ * https://issues.apache.org/jira/browse/HIVE-13632
+ */
+public class TestDataWritableWriter
+{
+    private static final Logger log = Logger.get(DataWritableWriter.class);
+    private final RecordConsumer recordConsumer;
+    private final GroupType schema;
+
+    public TestDataWritableWriter(final RecordConsumer recordConsumer, final GroupType schema)
```

Comment:
remove `final`

Reply:
Same as above, this is a class from apache/parquet-mr (Hive Parquet writer, which Presto uses to write parquet data https://github.com/prestodb/presto-hive-apache/blob/master/pom.xml)
There is a comment on the top of this class, that I applied patch for empty arrays/maps from https://issues.apache.org/jira/browse/HIVE-13632
Maybe it's better to leave this class's source close to the original version, and consider updating dependency in  https://github.com/prestodb/presto-hive-apache to Hive Parquet writer with fix from HIVE-13632.

Reply:
@dain when we copy files from other open source projects, do we want to update them to match our conventions/guidelines/etc. or do we just copy them as is?